File: pkg/dev_compiler/test/codegen/notnull.dart
// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library notnull;

void intAssignments() {
  var i = 0;
  i &= 1;
  i |= 1;
  i ^= 1;
  i >>= 1;
  i <<= 1;
  i -= 1;
  i %= 1;
  i += 1;
  i ??= 1;
  i *= 1;
  i ~/= 1;
  i++;
  --i;
  print(i + 1);

  int j = 1;
  j = i < 10 ? 1 : 2;
  print(j + 1);
}

void doubleAssignments() {
  var d = 0.0;
  d /= 1;
  print(d + 1);
}

void boolAssignments() {
  var b = true;
  b != b;
  print(b);
}

void increments() {
  int i = 1;
  print(++i);
  print(i++);
  print(--i);
  print(i--);

  int j;
  j = 1;
  print(++j);
  print(j++);
  print(--j);
  print(j--);
}

void conditionals([bool cond]) {
  int nullable;
  nullable = 1;
  int nonNullable = 1;
  int a = cond ? nullable : nullable;
  int b = cond ? nullable : nonNullable;
  int c = cond ? nonNullable : nonNullable;
  int d = cond ? nonNullable : nullable;
  print(a + b + c + d);
}

void nullAwareOps() {
  int nullable;
  int nonNullable = 1;
  int a = nullable ?? nullable;
  int b = nullable ?? nonNullable;
  int c = nonNullable ?? nonNullable;
  int d = nonNullable ?? nullable;
  print(a + b + c + d);

  var s = "";
  print(s?.length + 1);
}

void nullableLocals(int param) {
  print(param + 1);

  int i;
  // We could detect that i is effectively non-nullable with flow analysis.
  i = 1;
  print(i + 1);

  int j = 1;
  j = i == 1 ? 1 : null;
  print(j + 1);
}

void optParams([int x, int y = 1]) {
  print(x + y);
}

void namedParams({int x, int y : 1}) {
  print(x + y);
}

void forLoops(int length()) {
  for (int i = 0; i < 10; i++) {
    print(i + 1);
  }
  for (int i = 0; i < length(); i++) {
    print(i + 1);
  }
  for (int i = 0, n = length(); i < n; i++) {
    print(i + 1);
  }
  // TODO(ochafik): Special-case `int + 0` to provide a cheap way to coerce
  // ints to notnull in the SDK (like asm.js's `x|0` pattern).
  for (int i = 0, n = length() + 0; i < n; i++) {
    print(i + 1);
  }
}

void nullableCycle() {
  int x = 1;
  int y = 2;
  int z;
  x = y;
  y = z;
  z = x;
  print(x + y + z);

  int s;
  s = s;
  print(s + 1);
}

void nonNullableCycle() {
  int x = 1;
  int y = 2;
  int z = 3;
  x = y;
  y = z;
  z = x;
  print(x + y + z);

  int s = 1;
  s = s;
  print(s + 1);
}

class Foo {
  int intField;
  var varField;
  f(Foo o) {
    print(1 + varField + 2);
    while (varField < 10) varField++;
    while (varField < 10) varField = varField + 1;

    print(1 + intField + 2);
    while (intField < 10) intField++;
    while (intField < 10) intField = intField + 1;

    print(1 + o.intField + 2);
    while (o.intField < 10) o.intField++;
    while (o.intField < 10) o.intField = o.intField + 1;
  }
}

int _foo() => 1;
calls() {
  int a = 1;
  int b = 1;
  b = ((x) => x)(a);
  print(b + 1);

  int c = _foo();
  print(c + 1);
}

localEscapes() {
  int a = 1;
  var f = (x) => a = x;

  int b = 1;
  g(x) => b = x;

  f(1);
  g(1);

  print(a + b);
}

controlFlow() {
  for (int i, j;;) {
    i = j = 1;
    print(i + j + 1);
    break;
  }
  try {
    throw 1;
  } catch (e) {
    print(e + 1);
  }
  try {
    (null as dynamic).foo();
  } catch (e, trace) {
    print('${(e is String) ? e : e.toString()} at $trace');
  }
}

main() {
  intAssignments();
  doubleAssignments();
  boolAssignments();
  nullableLocals(1);
  optParams(1, 2);
  namedParams(x: 1, y: 2);
  forLoops(() => 10);
  increments();
  conditionals(true);
  calls();
  localEscapes();
  controlFlow();

  nullableCycle();
  nonNullableCycle();
}

File: pkg/dev_compiler/test/codegen/expect/notnull.js
dart_library.library('notnull', null, /* Imports */[
  'dart_sdk'
], function(exports, dart_sdk) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const notnull = Object.create(null);
  notnull.intAssignments = function() {
    let i = 0;
    i = i & 1;
    i = (i | 1) >>> 0;
    i = (i ^ 1) >>> 0;
    i = i[dartx['>>']](1);
    i = i << 1 >>> 0;
    i = i - 1;
    i = i[dartx['%']](1);
    i = i + 1;
    let t = i;
    t == null ? i = 1 : t;
    i = i * 1;
    i = (i / 1)[dartx.truncate]();
    i++;
    --i;
    core.print(i + 1);
    let j = 1;
    j = i < 10 ? 1 : 2;
    core.print(j + 1);
  };
  dart.fn(notnull.intAssignments, dart.void, []);
  notnull.doubleAssignments = function() {
    let d = 0.0;
    d = d / 1;
    core.print(d + 1);
  };
  dart.fn(notnull.doubleAssignments, dart.void, []);
  notnull.boolAssignments = function() {
    let b = true;
    b != b;
    core.print(b);
  };
  dart.fn(notnull.boolAssignments, dart.void, []);
  notnull.increments = function() {
    let i = 1;
    core.print(++i);
    core.print(i++);
    core.print(--i);
    core.print(i--);
    let j = null;
    j = 1;
    core.print((j = dart.notNull(j) + 1));
    core.print((() => {
      let x = j;
      j = dart.notNull(x) + 1;
      return x;
    })());
    core.print((j = dart.notNull(j) - 1));
    core.print((() => {
      let x = j;
      j = dart.notNull(x) - 1;
      return x;
    })());
  };
  dart.fn(notnull.increments, dart.void, []);
  notnull.conditionals = function(cond) {
    if (cond === void 0) cond = null;
    let nullable = null;
    nullable = 1;
    let nonNullable = 1;
    let a = dart.notNull(cond) ? nullable : nullable;
    let b = dart.notNull(cond) ? nullable : nonNullable;
    let c = dart.notNull(cond) ? nonNullable : nonNullable;
    let d = dart.notNull(cond) ? nonNullable : nullable;
    core.print(dart.notNull(a) + dart.notNull(b) + c + dart.notNull(d));
  };
  dart.fn(notnull.conditionals, dart.void, [], [core.bool]);
  notnull.nullAwareOps = function() {
    let nullable = null;
    let nonNullable = 1;
    let a = (nullable != null ? nullable : nullable);
    let b = (nullable != null ? nullable : nonNullable);
    let c = nonNullable;
    let d = nonNullable;
    core.print(dart.notNull(a) + dart.notNull(b) + c + d);
    let s = "";
    core.print(dart.notNull(s[dartx.length]) + 1);
  };
  dart.fn(notnull.nullAwareOps, dart.void, []);
  notnull.nullableLocals = function(param) {
    core.print(dart.notNull(param) + 1);
    let i = null;
    i = 1;
    core.print(dart.notNull(i) + 1);
    let j = 1;
    j = i == 1 ? 1 : null;
    core.print(dart.notNull(j) + 1);
  };
  dart.fn(notnull.nullableLocals, dart.void, [core.int]);
  notnull.optParams = function(x, y) {
    if (x === void 0) x = null;
    if (y === void 0) y = 1;
    core.print(dart.notNull(x) + dart.notNull(y));
  };
  dart.fn(notnull.optParams, dart.void, [], [core.int, core.int]);
  notnull.namedParams = function(opts) {
    let x = opts && 'x' in opts ? opts.x : null;
    let y = opts && 'y' in opts ? opts.y : 1;
    core.print(dart.notNull(x) + dart.notNull(y));
  };
  dart.fn(notnull.namedParams, dart.void, [], {x: core.int, y: core.int});
  notnull.forLoops = function(length) {
    for (let i = 0; i < 10; i++) {
      core.print(i + 1);
    }
    for (let i = 0; i < dart.notNull(length()); i++) {
      core.print(i + 1);
    }
    for (let i = 0, n = length(); i < dart.notNull(n); i++) {
      core.print(i + 1);
    }
    for (let i = 0, n = dart.notNull(length()) + 0; i < n; i++) {
      core.print(i + 1);
    }
  };
  dart.fn(notnull.forLoops, dart.void, [dart.functionType(core.int, [])]);
  notnull.nullableCycle = function() {
    let x = 1;
    let y = 2;
    let z = null;
    x = y;
    y = z;
    z = x;
    core.print(dart.notNull(x) + dart.notNull(y) + dart.notNull(z));
    let s = null;
    s = s;
    core.print(dart.notNull(s) + 1);
  };
  dart.fn(notnull.nullableCycle, dart.void, []);
  notnull.nonNullableCycle = function() {
    let x = 1;
    let y = 2;
    let z = 3;
    x = y;
    y = z;
    z = x;
    core.print(x + y + z);
    let s = 1;
    s = s;
    core.print(s + 1);
  };
  dart.fn(notnull.nonNullableCycle, dart.void, []);
  notnull.Foo = class Foo extends core.Object {
    Foo() {
      this.intField = null;
      this.varField = null;
    }
    f(o) {
      core.print(1 + dart.notNull(dart.as(this.varField, core.num)) + 2);
      while (dart.notNull(dart.as(dart.dsend(this.varField, '<', 10), core.bool))) {
        this.varField = dart.dsend(this.varField, '+', 1);
      }
      while (dart.notNull(dart.as(dart.dsend(this.varField, '<', 10), core.bool)))
        this.varField = dart.dsend(this.varField, '+', 1);
      core.print(1 + dart.notNull(this.intField) + 2);
      while (dart.notNull(this.intField) < 10) {
        this.intField = dart.notNull(this.intField) + 1;
      }
      while (dart.notNull(this.intField) < 10)
        this.intField = dart.notNull(this.intField) + 1;
      core.print(1 + dart.notNull(o.intField) + 2);
      while (dart.notNull(o.intField) < 10) {
        o.intField = dart.notNull(o.intField) + 1;
      }
      while (dart.notNull(o.intField) < 10)
        o.intField = dart.notNull(o.intField) + 1;
    }
  };
  dart.setSignature(notnull.Foo, {
    methods: () => ({f: [dart.dynamic, [notnull.Foo]]})
  });
  notnull._foo = function() {
    return 1;
  };
  dart.fn(notnull._foo, core.int, []);
  notnull.calls = function() {
    let a = 1;
    let b = 1;
    b = dart.as(dart.dcall(dart.fn(x => x), a), core.int);
    core.print(dart.notNull(b) + 1);
    let c = notnull._foo();
    core.print(dart.notNull(c) + 1);
  };
  dart.fn(notnull.calls);
  notnull.localEscapes = function() {
    let a = 1;
    let f = dart.fn(x => a = dart.as(x, core.int));
    let b = 1;
    function g(x) {
      return b = dart.as(x, core.int);
    }
    dart.fn(g);
    dart.dcall(f, 1);
    g(1);
    core.print(dart.notNull(a) + dart.notNull(b));
  };
  dart.fn(notnull.localEscapes);
  notnull.controlFlow = function() {
    for (let i = null, j = null;;) {
      i = j = 1;
      core.print(dart.notNull(i) + dart.notNull(j) + 1);
      break;
    }
    try {
      dart.throw(1);
    } catch (e) {
      core.print(dart.dsend(e, '+', 1));
    }

    try {
      dart.dsend(null, 'foo');
    } catch (e) {
      let trace = dart.stackTrace(e);
      core.print(`${typeof e == 'string' ? e : dart.toString(e)} at ${trace}`);
    }

  };
  dart.fn(notnull.controlFlow);
  notnull.cascadesOnNull = function() {
    let x = null;
    core.print(dart.hashCode(((() => {
      dart.toString(x);
      dart.runtimeType(x);
      return x;
    })())));
    let y = null;
    core.print(dart.hashCode(((() => {
      dart.toString(y);
      dart.runtimeType(y);
      return y;
    })())));
  };
  dart.fn(notnull.cascadesOnNull);
  notnull.main = function() {
    notnull.intAssignments();
    notnull.doubleAssignments();
    notnull.boolAssignments();
    notnull.nullableLocals(1);
    notnull.optParams(1, 2);
    notnull.namedParams({x: 1, y: 2});
    notnull.forLoops(dart.fn(() => 10, core.int, []));
    notnull.increments();
    notnull.conditionals(true);
    notnull.calls();
    notnull.localEscapes();
    notnull.controlFlow();
    notnull.cascadesOnNull();
    notnull.nullableCycle();
    notnull.nonNullableCycle();
  };
  dart.fn(notnull.main);
  // Exports:
  exports.notnull = notnull;
});
